### ISSUE TITLE
Correlation Loss

### ISSUE THREAD

**Issue by redhat12345** [SSARCandy/DeepCORAL#5]
From my understanding, we know that tensorflow and pytorch can calculate gradients automatically. In original coral implementation, they used caffe which cannot calculate gradients automatically and thats why they used back-propagation. So can we use simply the following correlation loss in pytorch as well? (https://github.com/tensorflow/models/blob/master/research/domain_adaptation/domain_separation/losses.py)

```py
def correlation_loss(source_samples, target_samples):
    source_samples -= tf.reduce_mean(source_samples, 0)
    target_samples -= tf.reduce_mean(target_samples, 0)
    source_samples = tf.nn.l2_normalize(source_samples, 1)
    target_samples = tf.nn.l2_normalize(target_samples, 1)
    source_cov = tf.matmul(tf.transpose(source_samples), source_samples)
    target_cov = tf.matmul(tf.transpose(target_samples), target_samples)
    corr_loss = tf.reduce_mean(tf.square(source_cov - target_cov))
    return corr_loss
```

**Comment by SSARCandy** [comment]
in my implementation, I defined a custom loss function to do the CORAL forward and backward.
The code is in https://github.com/SSARCandy/DeepCORAL/blob/master/models.py#L27

Sorry that I'm not quite understand what is your suggestion..?

**Comment by redhat12345** [comment]
Oh Now I understand. Some functions of pytorch and tensorflow cannot calculate gradients automatically. That's why you need both forward and backward of CARAL layer. Thanks man. Great work. I hope you will get the desired results soon as the paper. I run the original implementation in caffe and I got the same results that they provided in their paper.

**Comment by redhat12345** [comment]
I define coral loss like this in pytorch:

```py
def DeepCoral(source, target):
    xm = torch.mean(source, 1, keepdim=True)
    xc = torch.matmul(torch.transpose(xm, 0, 1), xm)  # source covariance
    xmt = torch.mean(target, 1, keepdim=True)
    xct = torch.matmul(torch.transpose(xmt, 0, 1), xmt)   # target covariance
    loss = torch.mean(torch.mul((xc - xct), (xc - xct)))   # frobenius norm between source and target
    return loss
```

The functions I used in the loss function, in pytorch it can calculate gradients automatically. Am I right ? Can you use this loss in your method, please?

**Comment by deep0learning** [comment]
@redhat12345 

I think your defined DeepCORAL loss will also work.

**Comment by ZhangJUJU** [comment]
@redhat12345 
I run your code,but the loss become nan .I think you forget to divide the number of source and target,so when my source sample are large,the loss become nan.you'd better devide the matrix by source number,target number and d,like the code does.
Now I am continue to test your code and try to fix the bugs.I will report the result later.

**Comment by redhat12345** [comment]
@ZhangJUJU 
Thank you so much. Please let me know the update.

**Comment by deep0learning** [comment]
@ZhangJUJU 

You are right.

**Comment by deep0learning** [comment]
I don't know why the performance is low than original implementation but if you use resnet, you will get the desired accuracy. I used alexnet in my experiment and I got the low accuracy but when I used resnet I got good accuracy.

**Comment by redhat12345** [comment]
Did you try to use the above loss yet? Any update please?

**Comment by SSARCandy** [comment]
Hi, this is the result I get using your loss, the performance didn't get better.
![loss](https://user-images.githubusercontent.com/6803791/32849791-2e80c6b6-ca6b-11e7-97ac-bc1e8bcc52c7.png)
![accuracy](https://user-images.githubusercontent.com/6803791/32849792-2eae4bae-ca6b-11e7-8db6-0be8d42120bd.png)

I slightly modify your loss by divided loss with `4*d*d`:

```py
def DeepCoral(source, target):
    d = source.data.shape[1]
    xm = torch.mean(source, 1, keepdim=True)
    xc = torch.matmul(torch.transpose(xm, 0, 1), xm)  # source covariance
    xmt = torch.mean(target, 1, keepdim=True)
    xct = torch.matmul(torch.transpose(xmt, 0, 1), xmt)   # target covariance
    loss = torch.mean(torch.mul((xc - xct), (xc - xct)))   # frobenius norm between source and target
    return loss/(4*d*d)
```

**Comment by redhat12345** [comment]
Thank you so much. I have sent you an email on ssarcandy@gmail.com. Please reply.

**Comment by jdily** [comment]
Hi @redhat12345,
I tried your loss function and got the similar results as @SSARCandy plotted.
It seems like your Coral loss function did not help to improve the performance...
Did you ever tried it on your own?
And since you mentioned you ran the caffe implementation and get the same performance as original paper, can you share the plotted result? 

thx!

**Comment by redhat12345** [comment]
@jdily 
My coral loss function's result.

![image](https://user-images.githubusercontent.com/32477541/34189519-91d42d16-e587-11e7-932f-51684f720ec2.png)
